### PR TITLE
OpInfo: `zero_`

### PR DESCRIPTION
**Pull request by krshrimali**
See #54261

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:6116]: Does this work? AFAIK, there is no operator named `zero_` in torch namespace.
You might need to need wrap the operator in lambda.
https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/torch/testing/_internal/common_methods_invocations.py#L4383

File: torch/testing/_internal/common_methods_invocations.py
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:6118]: This should support autograd.
https://github.com/pytorch/pytorch/blob/6de11c2db37d2b64debb97af40147de1581f718a/tools/autograd/derivatives.yaml#L1274

File: torch/testing/_internal/common_methods_invocations.py
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:970]: I think it is okay merge these two. It won't be hard to read.

**Review by kshitij12345 — COMMENTED** [review]
Thanks @krshrimali .

Have asked a few questions.

File: torch/testing/_internal/common_methods_invocations.py
Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:6118]: This is the error I got for almost all the tests:

```
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

Do you have any idea what is the solution to this? @kshitij12345

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:6118]: Do you still get an error after wrapping the operator in lambda?

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:6118]: I think the tests always assume that the operator's function variant (which was always true except for this operator) doesn't modify the input inplace.

eager consistency test has some mechanism to deal with this,

https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/test/test_ops.py#L321

Need to dig a bit further.

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:6118]: REG: https://github.com/pytorch/pytorch/pull/58731#discussion_r636674476

Yes, still getting the same errors after wrapping the operator in lambda.

Comment (kshitij12345) [reply torch/testing/_internal/common_methods_invocations.py:6118]: Inspired by 
https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/test/test_ops.py#L321

I tried,
```python
OpInfo('zero',
       op=lambda x: torch.zero_(x.clone()),
       dtypes=all_types_and_complex_and(torch.bfloat16, torch.half, torch.bool),
       supports_autograd=True,
       supports_out=False,
       sample_inputs_func=sample_inputs_zero_),
```

which works for all test except `jit` and `eager` consistency.
`jit` consistency has issues with op being wrapped in lambda.

The eager test fails for more subtle reasons.

https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/torch/testing/_internal/common_methods_invocations.py#L224-L228

We get the `method` variant and `inplace` variant in the above way.

For this case, `method` variant is `Tensor.zero_` and inplace variant is `None`.

Thus in eager tests at the point below,
https://github.com/pytorch/pytorch/blob/d88d321ee3a65894baf6dc0a580b44180ec52f3c/test/test_ops.py#L321-L323

We incorrectly don't clone the inputs for method variant (which is actually the inplace variant) and hence we get,

```python
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```

To mitigate this, I propose to allow setting `method_variant` and `inplace_variant` via the OpInfo constructor.

cc: @mruberry

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:6118]: Thanks, @kshitij12345 for digging into this. Taking a look at `test/test_autograd.py` to understand how autograd is tested for `method_tests`, it seems like `requires_grad` is set to `False` (/reset) for inplace ops. There is an additional comment that mentions that it needs to be fixed. Please see the relevant snippet below.

```python
is_magic_method = name[:2] == '__' and name[-2] == '__'
is_inpalce = name[-1] == "_" and not is_magic_method
self_variable = create_input((self_size,), dtype=dtype, device=device)[0][0]
# FixMe: run grad checks on inplace self
if is_inplace:
    self_variable.requires_grad = False
```

Reference: https://github.com/pytorch/pytorch/blob/master/test/test_autograd.py#L5474

Comment (krshrimali) [reply torch/testing/_internal/common_methods_invocations.py:6118]: For the scope of this PR, I suggest to jit and eager consistency checks, and we can file this an issue (on allowing to set `method_variant` and `inplace_variant` via OpInfo constructor) separately and take the discussion there.

Looking forward to know your views, @kshitij12345, @mruberry.

**Comment by kshitij12345** [comment]
Also, `torch.zero_` is not documented. File an issue? cc: @mruberry 

I think marking it as function variant was not intentional?
Throughout the codebase, I just see two entries for `torch.zero_` that too in benchmark

Screenshot from 2021-05-21 20-22-10

while, `zero_` inplace method is used more frequently.
Screenshot from 2021-05-21 20-26-50

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:967]: Nit: this probably doesn't need the (S, S) case

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:6113]: Skipping those tests and following up by allowing method and inplace variants to be specified sounds good

**Review by mruberry — COMMENTED** [review]
Overall this looks but, per discussion, some follow-up is needed.

For now, actually, I'd keep "supports_autograd=False" and add a comment with a link to an issue for that not being correct, but our tests not correctly dealing with this case.

Then I'd file another issue for setting the method variant and inplace variants for operators directly, AND we should support ops that don't have function variants but only have method and/or inplace variants, too. Solving this latter issue should let us address the first issue.

How does that sound @krshrimali, @kshitij12345?

**Comment by kshitij12345** [comment]
> For now, actually, I'd keep "supports_autograd=False" and add a comment with a link to an issue for that not being correct, but our tests not correctly dealing with this case.

I think that is fine since even `method_tests` are not doing the grad tests.

> Then I'd file another issue for setting the method variant and inplace variants for operators directly, AND we should support ops that don't have function variants but only have method and/or inplace variants, too. Solving this latter issue should let us address the first issue.

On this, it actually has a function variant is which weird, `torch.zero_` which very rarely used. Should it be deprecated? It has very few actual usages across the code base and probably should be replaced with `tensor.zero_()`.

Screenshot from 2021-05-23 14-14-39
Screenshot from 2021-05-23 14-14-22

@mruberry what do you think?

**Comment by mruberry** [comment]
> > For now, actually, I'd keep "supports_autograd=False" and add a comment with a link to an issue for that not being correct, but our tests not correctly dealing with this case.
> 
> I think that is fine since even `method_tests` are not doing the grad tests.
> 
> > Then I'd file another issue for setting the method variant and inplace variants for operators directly, AND we should support ops that don't have function variants but only have method and/or inplace variants, too. Solving this latter issue should let us address the first issue.
> 
> On this, it actually has a function variant is which weird, `torch.zero_` which very rarely used. Should it be deprecated? It has very few actual usages across the code base and probably should be replaced with `tensor.zero_()`.
> 
> Screenshot from 2021-05-23 14-14-39
> Screenshot from 2021-05-23 14-14-22
> 
> @mruberry what do you think?

Unfortunately we're inconsistent today with how the inplace versions of operators are exposed. Your question is a good one but this is one discrepancy we probably don't want to worry about at the moment as it's dependent on some future decisions about how we represent operators.

**Comment by krshrimali** [comment]
**Updates:**

Adding `method_variant` and `inplace_variant` args to `OpInfo` class is being done here. I think this (https://github.com/pytorch/pytorch/pull/59138) PR can wait for a few days until the PR is merged (linked before). The jit issue will still be there though, and an issue will be raised on that soon. Thanks for the patience!

**Review by kshitij12345 — APPROVED** [review]
LGTM! Thanks!

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6558]: Nice use of the new architecture

**Review by mruberry — APPROVED** [review]
Nice work, @krshrimali!

**Pull request closed without merging**
